opt_einsum/paths.py

```diff
@@ -652,6 +655,131 @@ def greedy(inputs, output, size_dict, memory_limit=None, choose_fn=None, cost_fn
     return ssa_to_linear(ssa_path)
 
 
+def _tree_to_sequence(c):
+    # converts a contraction tree to a contraction sequence, e.g.
+    #
+    # ((1,2),(0,(4,3))) --> [(1,2), (1,2), (0,2), (0,1)]
+    # 
+    # 0     0         0         (1,2)     --> ((1,2),(0,(3,4)))
+    # 1     3     --> (1,2) --> (0,(3,4))
+    # 2 --> 4         (3,4)
+    # 3     (1,2)
+    # 4
+    # 
+    # this function iterates through the table shown above from right to left; 
+    # note that contractions are commutative, i.e. (j,k) = (k,j);
+    # note that the solution is not unique;
+    
+    c = [c] # list of remaining contractions (lower part of columns shown above)
+    t = []  # list of elementary tensors (upper part of colums)
+    s = []  # resulting contraction sequence
+    
+    while len(c) > 0:
+        j, k = c.pop(-1)
+        s.insert(0, tuple())
+        
+        for i in sorted([i for i in [j, k] if type(i) == int]):
+            s[0] += (sum(1 for q in t if q < i),)
+            t.insert(s[0][-1], i)
+        
+        for i in [i for i in [j, k] if type(i) != int]:
+            s[0] += (len(t) + len(c),)
+            c.append(i)
+    
+    return s
+
+
+class DynamicProgrammingOptimizer(PathOptimizer):
+    
+    def __call__(self, inputs, output, size_dict, memory_limit=None):
+        
+        union = lambda a: functools.reduce(lambda x, y: x | y, a, set())
+        prod = lambda a: functools.reduce(lambda x, y: x * y, a, 1)
```

Comment:
You could use ``helpers.compute_size_by_dict(inds, size_dict)`` directly instead of this.

Reply:
Okay, good to know. I replaced my calls to `prod`.